Sources/NIOCore/ByteBuffer-binaryEncodedLengthPrefix.swift

```diff
@@ -0,0 +1,310 @@
+    /// If the the full integer can be read, move the reader index to after the integer, and return the integer.
+    /// - Parameters:
+    ///   - as: The type of integer to be read.
+    ///   - buffer: The buffer to read from.
+    /// - Returns: The integer that was read, or nil if it was not possible to read it.
+    func readInteger<IntegerType: FixedWidthInteger>(
+        as: IntegerType.Type,
+        from buffer: inout ByteBuffer
+    ) -> IntegerType?
+
+    /// Write an integer to a buffer. Move the writer index to after the written integer.
+    /// - Parameters:
+    ///    - integer: The type of the integer to write.
+    ///    - buffer: The buffer to write to.
+    /// - Returns: The number of bytes used to write the integer.
+    func writeInteger<IntegerType: FixedWidthInteger>(
+        _ integer: IntegerType,
+        to buffer: inout ByteBuffer
+    ) -> Int
+
+    /// An estimate of the bytes required to write integers using ths strategy.
+    /// Callers may use this to reserve bytes before writing the integer.
+    /// If the actual bytes used by the write function is more or less than this, it may be necessary to shuffle bytes.
+    /// Therefore, an accurate prediction here will improve performance.
+    var requiredBytesHint: Int { get }
+
+    /// Write an integer to a buffer. Move the writer index to after the written integer.
+    /// Call this function if you have already reserved some capacity for an integer to be written.
+    /// Implementors should consider using a less efficient encoding, if possible,to fit exactly within the reserved capacity.
+    /// Otherwise, callers may need to shift bytes to reconcile the difference.
+    /// It is up to the implementor to find the balance between performance and size.
+    /// - Parameters:
+    ///   - integer: The integer to write
+    ///   - reservedCapacity: The capacity already reserved for writing this integer
+    ///   - buffer: The buffer to write into.
+    /// - Returns: The number of bytes used to write the integer.
+    func writeIntegerWithReservedCapacity(
+        _ integer: Int,
+        reservedCapacity: Int,
+        to buffer: inout ByteBuffer
+    ) -> Int
+}
+
+extension NIOBinaryIntegerEncodingStrategy {
+    @inlinable
+    public var requiredBytesHint: Int { 1 }
+
+    @inlinable
+    public func writeIntegerWithReservedCapacity<IntegerType: FixedWidthInteger>(
+        _ integer: IntegerType,
+        reservedCapacity: Int,
+        to buffer: inout ByteBuffer
+    ) -> Int {
+        self.writeInteger(integer, to: &buffer)
+    }
+}
+
+extension ByteBuffer {
+    /// Read a binary encoded integer, moving the `readerIndex` appropriately.
+    /// If there are not enough bytes, nil is returned.
+    @inlinable
+    public mutating func readEncodedInteger<Strategy: NIOBinaryIntegerEncodingStrategy>(_ strategy: Strategy) -> Int? {
```

Comment:
These should probably be generic over `FixedWidthInteger`, the way the existing `readInteger` and `writeInteger` functions are.